### PR TITLE
S3 caching

### PR DESCRIPTION
**Pull request by rodneykinney**
Introduce the option to turn off persistent caching of S3 artifacts.
@andrewlmurray Please review
@schmmd FYI

Also includes a minor bug fix related to the `runOnly` config
@markschaake We came across this bug before my vacation.

### CODE REVIEW

File: src/main/scala/org/allenai/pipeline/Pipeline.scala [hunk 1]
```diff
@@ -178,10 +180,11 @@ trait Pipeline extends Logging {
   def runOnly(title: String, targetNames: Iterable[String]): Iterable[(String, Any)] = {
     val targets = getStepsByName(targetNames)
     val targetStepInfo = targets.map(_._2.stepInfo).toSet
-    val allDependencies = targets.flatMap { case (s, p) => Workflow.upstreamDependencies(p) }
+    val directDependencies = targets.flatMap { case (s, p) => p.stepInfo.dependencies.map(_._2) }
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/Pipeline.scala:183]: Bug fix: runOnly will throw exception if upstream dependencies don't exist, but should only check direct upstream dependencies

File: src/main/scala/org/allenai/pipeline/Pipeline.scala [hunk 2]
```diff
@@ -298,12 +301,14 @@ trait ConfiguredPipeline extends Pipeline {
       .getOrElse(new File(System.getProperty("user.dir")).toURI)
 
   protected[this] def getStringList(key: String) =
-    config.get[String](key) match {
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/Pipeline.scala:301]: Bug fix: more robust config parsing (accept either single string or string list)

File: src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala [hunk 3]
```diff
@@ -34,20 +36,16 @@ class S3FlatArtifact(
   val path: String,
   val config: S3Config,
   val contentTypeOverride: Option[String] = None
-)
-    extends FlatArtifact with S3Artifact[FileArtifact] {
-  protected def makeLocalArtifact(f: File) = new FileArtifact(f)
+)(implicit cachingImpl: S3Cache = DefaultS3Cache)
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala:39]: Introduce an S3Cache class that manages reading/writing of artifacts in S3

File: src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala [hunk 4]
```diff
@@ -122,45 +115,110 @@ trait S3Artifact[A <: Artifact] extends Logging {
...
+
+object DefaultS3Cache extends LocalS3Cache()
+
+/** Caches S3 artifacts on the local filesystem
+  * @param persistentCacheDir If specified, cached objects will be stored in a directory
+  *                           that is persistent across JVM processes. There are no limits
+  *                           on the size of the data stored in the persistent directory
+  */
+case class LocalS3Cache(
+    persistentCacheDir: Option[File] = Some(new File(System.getProperty("java.io.tmpdir"), "pipeline-cache"))
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala:153]: Default behavior is unchanged from before

File: src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala [hunk 5]
```diff
@@ -122,45 +115,110 @@ trait S3Artifact[A <: Artifact] extends Logging {
...
+) extends S3Cache with Logging {
+
+  override def readFlat(artifact: S3Artifact) = {
+    if (usePersistentCache) {
+      new FileArtifact(downloadLocalCopy(artifact)).read
+    } else {
+      artifact.readContents()
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala:160]: Read directly from S3 if no persistent cache directory specified.

File: src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala [hunk 6]
```diff
@@ -122,45 +115,110 @@ trait S3Artifact[A <: Artifact] extends Logging {
...
+    artifact.upload(local)
+    result
+  }
+
+  override def readZip(artifact: S3Artifact): Reader = {
+    new ZipFileArtifact(downloadLocalCopy(artifact)).reader
+  }
+
+  override def writeZip[T](artifact: S3Artifact, writer: Writer => T) = {
+    val local = createLocalCopy(artifact)
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala:178]: Zip files still have to be downloaded locally, but will be deleted on JVM exit if no persistent cache directory is specified.

File: src/main/scala/org/allenai/pipeline/s3/S3Pipeline.scala [hunk 7]
```diff
@@ -11,6 +11,7 @@ import java.net.URI
 
 trait S3Pipeline extends Pipeline {
   def credentials: S3Credentials = S3Config.environmentCredentials()
+  implicit val s3Cache: S3Cache = DefaultS3Cache
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/s3/S3Pipeline.scala:14]: To turn off persistent caching, override this val with `LocalS3Cache(None)` in the top-level pipeline class

File: src/test/scala/org/allenai/pipeline/TestPipeline.scala [hunk 8]
```diff
@@ -34,6 +34,17 @@ class TestPipeline extends UnitSpec with ScratchDirectory {
     p1File should exist
     p2File should not(exist)
 
+    {
+      // runOnly should succeed if direct dependencies exist,
+      // even if other upstream dependencies don't
+      val first = pipeline.Persist.Singleton.asText(p1, "First")
```
Comment (rodneykinney) [inline src/test/scala/org/allenai/pipeline/TestPipeline.scala:40]: Unit test for bug fix

**Comment by schmmd** [comment]
@andrewlmurray can we merge this down?

**Comment by andrewlmurray** [comment]
@rodneykinney LGTM. Just to check my understanding: it looks like this gets us part of the way home. 
zip files will still leak, just not across runs unless we use a persistent cache dir. Is this accurate?

**Comment by rodneykinney** [comment]
That's correct. We don't use zip files when persisting RDDs, so this should
effectively solve the problem.

**Pull request merged**